docs/modules/ROOT/pages/index.adoc

```diff
@@ -331,6 +342,59 @@ The `signer_payload` data structure MUST be presented to be signed by the _<<_cr
 
 IMPORTANT: If a data hash assertion is being used, the _<<C2PA claim generator>>_ MUST also follow the process described in xref:_interaction_with_data_hash_assertion[xrefstyle=full].
 
+==== Calculating the `expected_claim` value
+
+If the *<<_identity_assertion,identity assertion>> generator* chooses to include the `expected_claim` value in `signer_payload`, it MUST be computed as follows:
+
+. Create a valid _<<C2PA claim>>_ data structure that includes completed references to all _<<_c2pa_assertion,C2PA assertions>>,_ *including* all *<<_identity_assertion,identity assertions>>.*
+. For the specific *<<_identity_assertion,identity assertion>>* being computed, replace the `hash` value with all zero-value (`0x00`) bytes.
+. For any *<<_identity_assertion,identity assertions>>* whose signing credentials are described via the `expected_countersigners` field, replace the `hash` value with all zero-value (`0x00`) bytes.
```

Comment:
Line 349. Assuming the current identity assertion is the first under consideration, how can the yet to be signed identity assertion be represented in the claim?
Line 350. OK for the current identity assertion under consideration
line 351. Is this meant to address those other identity assertions which are using an expected claim field, i.e., lets substitute their yet-to-be-determined values with zeroes.? 

Replacing both earlier signed and later signed claim hash representations seems to lose out to an easy improvement in which later-signed also include earlier-signed (but only zeroes of the later-signed and themselves), as I think was in Paul's proposal (iirc). The order would have to be evident (from the claim structure?). However, earlier-signed (with only zeros as references to the later-signed) could still be removed and added as earlier-signed id assertions in an alternative manifest to the same content with changed later-signed identity referenced assertions (and then re-signed by the same c2pa cert)

I still think it would be complete if each identity assertion were to witness the hash of every other identity assertion's referenced assertion structure, in addition to the claim box structure, to prevent this weakness to early-signed identity assertions in the multiple identity assertion case.

Reply:
@nigelcearnshaw I'm tweaking the wording of lines 349-351 to make it clearer that lines 350 and 351 are exceptional cases that are exempted from the "completed references" clause in line 349.

> I still think it would be complete if each identity assertion were to witness the hash of every other identity assertion's referenced assertion structure

I think this isn't feasible for the same reason that we're doing the zeroing-out of hash values: Any computation that depends on the value of a `hash` reference to another assertion necessarily imposes an ordering to those assertions. The `referenced_assertions` structure can't be known until all _other_ assertions in that data structure are completed (including their signatures), which means there can't be any cycle of such references.

Please let me know if I'm missing something here.

Reply:
Thanks Eric,

Let me try this. 

I understand that the claims as seen by the C2PA claim generator are hashes of the jumbf, cbor encoded assertions. 
Since we cannot form the jumbf  cbor representation of the identity assertion until it is complete, including internal signature covering both its own internal contents and other claim map contents, there is a dependency on whether other identity assertions were complete before they could be included in the claim representation and this in turn would impose an order on creating the identity assertions. OK. So far I am in aggreement. 

I can also see that an identity assertion cannot reference its own claim representation, so this needs to be excluded when considering the manifest claim box, which is done here by zeroing out its entry. Again, OK with this.

What I am trying to explore is whether there is a better way to include the *contents* (not including signature) of the other not-yet-signed or already-signed  identity assertions as an alternative representation when completing the signature of any identity assertion. 

We have identified two ways already.

1. Including the final identity assertion claim representation for those that have already been signed and serialised, but zeroing out later-signed ones. This imposes an order and I acknowledge those concerns. It is also not perfect as a solution. 
2. Zeroing out all claim map entries as place holders for the correct values. This solves the ordering problem, but there ramains a vulnerability.  Each identity assertion has not witnessed the contents of the other identity assertions. This leaves open the loop-hole that an identity assertion can be removed from the current manifest and added to a separate manifest, with all of the same content BMFF assertions and claims (which,  I admit, is a constraint), but with changes to the other identity assertions (specifically a changed referenced assertions list) before that new replacement identity assertion is signed and that new manifest is then also signed.  I imagine this can be done on the same machine at a different time.

So while I can see that this arrangement severely constrains the scope of attack, it does not perfectly counter all possibilities.  Adversaries like possibilities.

A solution to this (I *think*) would be that rather than using zeroes as the replacement subsitute value for the hash of the identity assertion in the claim map when evaluating the payload of the identity assertion signature, a hash of the contents of the identity assertion (just the hash of the reference claims, not the signed hash) would be included. I don't think this would re-introduce the ordering issue as the contents of each identity assertion (not including signature)  are known independently of them being signed. 

This would now mean it is impossible to for the identity assertion to be valid if any of the assertions or claims change, including the contents of the accompanying identity assertions. 

For example, 

A publisher  publishes some content that includes a contribution of a poem from a poet. They include an identity assertion for the publisher, signed by the publisher and in order to promote the poet include an additional identity assertion provided by the poet and signed by the poet.  This is signed by a well known industry model of editor which is c2pa compliant and composes and signs the manifest (using zeroing out) using its c2pa certificate. 

However, later, another person with access to the same editor (or model with the same signing key) takes the manifest and replaces the poet's credentials with their own, adjusting the reference claims accordingly. As this changes the final claim entry for the manifest, the manifest has to be re-signed with the same or equivalent editor (to maintain the same signing key). This valid manifest is then distributed. 

This leaves the publisher in a very embarrassed position, as it now appears they have miss-attributed their contributing talent.  Their included identity assertion is still valid since when evaluating the signature, the offending replaced identity assertion claim value has been zeroed out.

Another example would be the owner of one of the identity assertions simply widening their scope of participation (changing the list of reference claims) after the event. This does depend on them having access to the same machine (or same model with signing key).

So that is my concern.  Look forward to your thoughts.

Reply:
[ Thinking again, this does work if identity assertions cannot have other identity assertions in their referenced assertions (seems reasonable) and if when hashing the other identity reference assertions map we do not include the 'expected claim' field that was calculated for that assertion. ]

Thanks Eric,

Let me try this. 

I understand that the claims as seen by the C2PA claim generator are hashes of the jumbf, cbor encoded assertions. 
Since we cannot form the jumbf  cbor representation of the identity assertion until it is complete, including internal signature covering both its own internal contents and other claim map contents, there is a dependency on whether other identity assertions were complete before they could be included in the claim representation and this in turn would impose an order on creating the identity assertions. OK. So far I am in aggreement. 

I can also see that an identity assertion cannot reference its own claim representation, so this needs to be excluded when considering the manifest claim box, which is done here by zeroing out its entry. Again, OK with this.

What I am trying to explore is whether there is a better way to include the *contents* (not including signature) of the other not-yet-signed or already-signed  identity assertions as an alternative representation when completing the signature of any identity assertion. 

We have identified two ways already.

1. Including the final identity assertion claim representation for those that have already been signed and serialised, but zeroing out later-signed ones. This imposes an order and I acknowledge those concerns. It is also not perfect as a solution. 
2. Zeroing out all claim map entries as place holders for the correct values. This solves the ordering problem, but there ramains a vulnerability.  Each identity assertion has not witnessed the contents of the other identity assertions. This leaves open the loop-hole that an identity assertion can be removed from the current manifest and added to a separate manifest, with all of the same content BMFF assertions and claims (which,  I admit, is a constraint), but with changes to the other identity assertions (specifically a changed referenced assertions list) before that new replacement identity assertion is signed and that new manifest is then also signed.  I imagine this can be done on the same machine at a different time.

So while I can see that this arrangement severely constrains the scope of attack, it does not perfectly counter all possibilities.  Adversaries like possibilities.

A solution to this (I *think*) would be that rather than using zeroes as the replacement subsitute value for the hash of the identity assertion in the claim map when evaluating the payload of the identity assertion signature, a hash of the contents of the identity assertion (just the hash of the reference claims, not the signed hash) would be included. I don't think this would re-introduce the ordering issue as the contents of each identity assertion (not including signature)  are known independently of them being signed. 

This would now mean it is impossible to for the identity assertion to be valid if any of the assertions or claims change, including the contents of the accompanying identity assertions. 

For example, 

A publisher  publishes some content that includes a contribution of a poem from a poet. They include an identity assertion for the publisher, signed by the publisher and in order to promote the poet include an additional identity assertion provided by the poet and signed by the poet.  This is signed by a well known industry model of editor which is c2pa compliant and composes and signs the manifest (using zeroing out) using its c2pa certificate. 

However, later, another person with access to the same editor (or model with the same signing key) takes the manifest and replaces the poet's credentials with their own, adjusting the reference claims accordingly. As this changes the final claim entry for the manifest, the manifest has to be re-signed with the same or equivalent editor (to maintain the same signing key). This valid manifest is then distributed. 

This leaves the publisher in a very embarrassed position, as it now appears they have miss-attributed their contributing talent.  Their included identity assertion is still valid since when evaluating the signature, the offending replaced identity assertion has been zeroed out.

Another example would be the owner of one of the identity assertions simply widening their scope of participation (changing the list of reference claims) after the event. This does depend on them having access to the same machine (or same model with signing key).

So that is my concern. 

Look forward to your thoughts.

Reply:
Future ID assertions will likely reference past ID assertions, so I don't think we know the hash of the future assertion at the time of signing the earlier assertion

Reply:
re:Future ID assertions will likely reference past ID assertions, so I don't think we know the hash of the future assertion at the time of signing the earlier assertion

But within a given manifest signing event, everything will be known?

Adding another future identity assertion will invalidate all the signatures including the current ID assertion signatures since even with zeroing out there will be a new zeroed-out entry in the claims to consider when evaluating  the 'expected claim' process.

Reply:
@nigelcearnshaw As agreed in today's CAWG meeting, we will address this issue in a separate PR to follow. See #114.